src/appimage.js

```diff
@@ -0,0 +1,84 @@
+import { debug, getInput, info, setFailed } from '@actions/core';
+import { chmod } from 'fs/promises';
+import { Options } from './options';
+import { downloadFile, getArchitecture } from './util';
+import { downloadLatestArtifact } from './workflow';
+
+/**
+ * @param {WorkingDirectory} workingDirectory
+ * @returns {Promise<void>}
+ */
+export async function downloadAppImage(workingDirectory) {
+    const version = getInput(Options.PackSquashVersion);
+    debug(`PackSquash version input variable value: ${version}`);
+    const arch = getArchitecture();
+    debug(`Runner architecture: ${arch}`);
+
+    let arch_infix;
+    switch (version) {
+        case 'v0.1.0':
+        case 'v0.1.1':
+        case 'v0.1.2':
+        case 'v0.2.0':
+        case 'v0.2.1':
+        case 'v0.3.0-rc.1':
+        case 'v0.3.0':
+            setFailed(`Unsupported PackSquash version: ${version}. Please use PackSquash-action@v2 instead.`);
+            break;
+        case 'latest':
+            switch (arch) {
+                case 'X86':
+                case 'X64':
+                    arch_infix = 'x64';
+                    break;
+                case 'ARM':
+                case 'ARM64':
+                    arch_infix = 'arm64';
+                    break;
+                default:
+                    setFailed(`The latest PackSquash build does not support ${arch}. Please use a runner with a supported architecture, or request support for it.`);
+            }
+            await downloadLatestAppImage(workingDirectory, arch_infix);
+            break;
+        case 'v0.3.1':
+            switch (arch) {
+                case 'X86':
+                case 'X64':
+                    arch_infix = 'x86_64';
+                    break;
+                case 'ARM':
+                case 'ARM64':
+                    arch_infix = 'aarch64';
+                    break;
+                default:
+                    setFailed(
+                        `PackSquash ${version} does not support ${arch}. Please use a runner with a supported architecture, or request support for it. A newer version might add support for this architecture.`
+                    );
+            }
+            await downloadReleaseAppImage(workingDirectory, version, `PackSquash-${version}-${arch_infix}.AppImage`);
+            break;
```

Comment:
I've two thoughts about these lines of code:

- 32-bit architectures (x86, ARM) can't run 64-bit binaries (x64, ARM64/AArch64) without resorting to emulation, but that is out of the scope for the action. 64-bit architectures and kernels are backwards compatible with 32-bit binaries, but those already have first-class support with native 64-bit binaries. Therefore, I think we should consider `X86` and `ARM` as unsupported because the 64-bit PackSquash binaries won't run on those architectures.
- The `downloadReleaseAppImage` and `downloadLatestAppImage` functions fail for unsupported architectures because `arch_infix` is never set (remember that `setFailed` does not abort the script execution). This causes the exported `downloadAppImage` to fail too. However, the download is attempted even if the architecture is unknown, which is not very nice. I think the `default` cases in this `switch` should throw exceptions instead of calling `setFailed`, to let the `index.js` exception handle call `setFailed` and exit.

Reply:
Fix by 15d6c8038cb22e5be974104021164c6963a59e25, 6c4fadf1a85e489ff7b46c0150b07c21a57ac8d0